juju/juju
# Inform user of storage destruction/detachment in remove-{application,unit}

## Description of change

We introduce new DestroyApplication and DestroyUnit
methods to the Application facade. These supersede
the existing Destroy and DestroyUnits methods,
respectively. The new methods are bulk, and return
information relating to the entities being destroyed.

This enables us to inform the user of storage that is
destroyed, or detached and left in the model, when
removing an application or unit with the remove-application
or remove-unit commands.

We will do the same for remove-machine in a follow-up.

Also, move a few tests over to the unit test suite in
apiserver/application.

## QA steps

1. juju bootstrap localhost
2. juju deploy postgresql -n2 --storage pgdata=rootfs
3. juju remove-unit postgresql/0
should see: 
> removing unit postgresql/0
> - removing storage pgdata/0
4. juju remove-application postgresql/1 --verbose
should see: 
> removing application postgresql
> - removing unit postgresql/1
> - removing storage pgdata/1

Also tested backwards compatibility (new server & old client, and vice versa).

## Documentation changes

- The remove-application command now accepts multiple application names, and will remove them all.
- The user will be informed of removals. When removing an application, the units are only listed if the user specified --verbose (as removing an application implies the units will *all* be removed, it's probably a bit noisy to be on by default)

## Bug reference

None.
